master/buildbot/reporters/slack.py

```diff
@@ -0,0 +1,177 @@
+    # pylint: disable=invalid-name, too-many-ancestors
+    """
+    minimal slack integration
+    """
+    def checkConfig(self,
+                    username,   # pylint: disable=unused-argument
+                    url,        # pylint: disable=unused-argument
+                    mapping,
+                    default_channels=_Default,
+                    *args, **kwargs):
+        """
+        check if the provided configuration is OK
+        """
+        if not isinstance(mapping, dict):
+            config.error('mapping parameter must be a dictionary')
+        if not (default_channels is _Default or
+                isinstance(default_channels, (list, tuple))):
+            config.error('default_channels parameter must be a list or a tuple')
+
+        super(SlackReporter).checkConfig(*args, **kwargs)
+
+    @defer.inlineCallbacks
+    def reconfigService(self, username, url, mapping, default_channels,
+                        *args, **kwargs):
+        """
+        reconfigure the service based on the new parameters
+        """
+        yield super(SlackReporter).reconfigService(self, *args, **kwargs)
+
+        # pylint: disable=attribute-defined-outside-init
+
+        self._username = username
+
+        # TODO: this goes to the documentation
+        # mapping is expected to have the following structure
+        # mapping = {
+        #     'builder1': {
+        #         'kinds': ('start',),
+        #         'channels': ('#channel1', '#channel2')
+        #     },
+        #     'builder2': {
+        #         'kinds': ('finish',),
+        #     },
+        #     'builder3': {
+        #         'kinds': ('start', 'finish')
+        #     },
+        # }
+        self._mapping = mapping
+        # TODO: this goes to the documentation
+        # list of channels to post a message to if none are specified in the
+        # mapping.  the default behaviour is to use the channel configured for
+        # the webhook
+        self._default_channels = default_channels
+
+        self._http = yield HTTPClientService.getService(self.master, url)
+
+    def _filter_builds(self, build):
+        """
+        check if we should post a message for the given build:
+
+        * the builder is present in the map
+        * 'kinds' has the value for the kind of this notification (start/finish)
+        """
+        not_interested = None, None
+
+        props = self._mapping.get(build['builder'], dict())
+        if not props:
+            return not_interested
+
+        kind = 'finish' if build['complete'] else 'start'
+        if kind not in props['kinds']:
+            return not_interested
+
+        if self._default_channels is not _Default:
+            channels = props.get('channels', self._default_channels)
+            if not channels:
+                return not_interested
+        else:
+            channels = None
+
+        return props, channels
+
+    @staticmethod
+    def _prepare_attachment(build):
+        """
```

Comment:
I think we should split this method into several methods that could be used as customization endpoints, so that people can post the exact message they need.

Thus, they have to be public methods